WKNavigationAction should expose _WKHitTestResult

jeffmapple opened the pull request:
#### 624a8e752b7108e9a5f547a6dc9bc432f41c501f
<pre>
WKNavigationAction should expose _WKHitTestResult
<a href="https://bugs.webkit.org/show_bug.cgi?id=246539">https://bugs.webkit.org/show_bug.cgi?id=246539</a>
&lt;rdar://100526247&gt;

Reviewed by Alex Christensen.

Add a _hitTestResult property to WKNavigationAction. This property can be nil.

* Source/WebKit/Shared/NavigationActionData.cpp:
(WebKit::NavigationActionData::encode const):
Encode webHitTestResultData.

(WebKit::NavigationActionData::decode):
Decode webHitTestResultData.

* Source/WebKit/Shared/NavigationActionData.h:
Add an optional webHitTestResultData.

* Source/WebKit/UIProcess/API/APINavigationAction.h:
Add webHitTestResultData().

* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction _hitTestResult]):
Added.

* Source/WebKit/UIProcess/API/Cocoa/WKNavigationActionPrivate.h:
Added _hitTestResult property.

* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::setWebHitTestResultDataInNavigationActionDataIfNecessary):
Added. Sets navigationActionData.webHitTestResultData if there is associated mouse event data.

(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
Set navigationActionData.webHitTestResultData if necessary.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/DecidePolicyForNavigationAction.mm:
Add _hitTestResult tests to tests that generate mouse event data. The only data that&apos;s populated in
_hitTestResult is the elementBoundingBox, so verify that it&apos;s not empty.

Canonical link: <a href="https://commits.webkit.org/256038@main">https://commits.webkit.org/256038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/feca56207e02a7718d786c02d207f0b09becdbad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3583 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/104063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98403 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3632 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31785 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2604 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80799 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38183 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4175 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39944 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41897 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -984,6 +986,15 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
+        HitTestResult hitTestResult = m_frame->coreFrame()->eventHandler().hitTestResultAtPoint(mouseEventData->absoluteLocation, hitType);
+
+        WebKit::WebHitTestResultData webHitTestResultData(hitTestResult, false);
+        navigationActionData.webHitTestResultData = std::make_optional(webHitTestResultData);
```
Can this use WTFMove?

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -984,6 +986,15 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
+#if PLATFORM(MAC) || HAVE(UIKIT_WITH_MOUSE_SUPPORT)
+    if (auto mouseEventData = navigationAction.mouseEventData()) {
+        constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::DisallowUserAgentShadowContent, HitTestRequest::Type::AllowChildFrameContent };
+        HitTestResult hitTestResult = m_frame->coreFrame()->eventHandler().hitTestResultAtPoint(mouseEventData->absoluteLocation, hitType);
```
This should go below the null check on line 1000

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -984,6 +986,15 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
     navigationActionData.lockBackForwardList = navigationAction.lockBackForwardList();
     navigationActionData.privateClickMeasurement = navigationAction.privateClickMeasurement();
+#if PLATFORM(MAC) || HAVE(UIKIT_WITH_MOUSE_SUPPORT)
+    if (auto mouseEventData = navigationAction.mouseEventData()) {
```
Do we need this in WebFrameLoaderClient::dispatchDecidePolicyForNewWindowAction too?

jeffmapple replied:
Good catch, yes we do.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
```diff
@@ -237,4 +239,18 @@ - (void)_storeSKAdNetworkAttribution
+    if (!webHitTestResultData)
+        return nil;
+
+    auto apiHitTestResult = API::HitTestResult::create(webHitTestResultData.value());
```
This does 2 unnecessary copies.  It can easily be reduced to 1 in this PR.

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -844,6 +846,21 @@ void WebFrameLoaderClient::dispatchDecidePolicyForResponse(const ResourceRespons
 }
 
+#if PLATFORM(MAC) || HAVE(UIKIT_WITH_MOUSE_SUPPORT)
+static void setWebHitTestResultDataInNavigationActionDataIfNecessary(const NavigationAction& navigationAction, NavigationActionData& navigationActionData, WebCore::Frame* coreFrame)
```
Parameter should be a Frame& if it's non-null, and it is in this case.

jeffmapple committed:
WKNavigationAction should expose _WKHitTestResult
https://bugs.webkit.org/show_bug.cgi?id=246539
<rdar://100526247>

Reviewed by Alex Christensen.

Add a _hitTestResult property to WKNavigationAction. This property can be nil.

* Source/WebKit/Shared/NavigationActionData.cpp:
(WebKit::NavigationActionData::encode const):
Encode webHitTestResultData.

(WebKit::NavigationActionData::decode):
Decode webHitTestResultData.

* Source/WebKit/Shared/NavigationActionData.h:
Add an optional webHitTestResultData.

* Source/WebKit/UIProcess/API/APINavigationAction.h:
Add webHitTestResultData().

* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction _hitTestResult]):
Added.

* Source/WebKit/UIProcess/API/Cocoa/WKNavigationActionPrivate.h:
Added _hitTestResult property.

* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::setWebHitTestResultDataInNavigationActionDataIfNecessary):
Added. Sets navigationActionData.webHitTestResultData if there is associated mouse event data.

(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
Set navigationActionData.webHitTestResultData if necessary.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/DecidePolicyForNavigationAction.mm:
Add _hitTestResult tests to tests that generate mouse event data. The only data that's populated in
_hitTestResult is the elementBoundingBox, so verify that it's not empty.

Canonical link: https://commits.webkit.org/256038@main
